Objects/StrategyCardAutomator.ttslua

```diff
@@ -1637,108 +1648,172 @@ function researchTech(color, cardName)
-	rot = techDeck.getRotation()
-
-	-- unit upgades are sent to factionSheet, other tech just put left of deck.
-	if overrides[cardName] and overrides[cardName]["upgrade"] then
-		-- code copied from setupHelper factionsheetsnappoints
-		local x0 = 1.09
-		local dx = (-1.09 * 2) / 3
-		local z0 = -0.74
-		local dz = (0.74 * 2) / 3
-
-		local counter = 1
-		local snapPoints = {}
-		for col = 0, 3 do
-			for row = 0, 3 do
-				if col <= row then
-					snapPoints[counter] = {
-							x = x0 + col * dx,
-							y = 0,
-							z = z0 + row * dz
-						}
-					counter = counter + 1
+	local guid = _getEntryGuid(techDeck, cardName)
+	if not guid then
+		_safeBroadcastToColor('Research Technology: ' .. cardName .. " not found in deck " .. techDeck.getName(), color, color)
+		return
+	end
+
+
+
+	local function _getPosToTechBoard(cardName, techBoard, techDeck)
+		local pos
+		local dx
+		local dz
+		-- find grid size and origin of snap points on tech board
+		for _,snap in ipairs(techBoard.getSnapPoints()) do
+			if not pos then
+				pos = snap.position
+			else
+				if (not dz) and (snap.position.z < (pos.z - 0.1)) then
+					dz = snap.position.z - pos.z
+				end
+				if (not dx) and (snap.position.x < (pos.x - 0.1)) then
+					dx = pos.x - snap.position.x
+				end
+				if snap.position.x > pos.x then
+					pos.x = snap.position.x
+				end
+				if snap.position.z > pos.z then
+					pos.z = snap.position.z
 				end
 			end
 		end
 
-		local pointorder = {
-		["Flagship"] = 1,
-		["War Sun"] = 2,
-		["Dreadnought"] = 3,
-		["Carrier"] = 4,
-		["Cruiser"] = 5,
-		["Destroyer"] = 6,
-		["Fighter"] = 7,
-		["PDS"] = 8,
-		["Infantry"] = 9,
-		["Space Dock"] = 10
-		}
+		-- find column for tech color
+		local _columns = {["Blue"]=0, ["Green"]=1, ["Yellow"]=2, ["Red"]=3, ["unitUpgrade"]=4}
+		column = _columns[_technologyHelper.getTechType(cardName)]
+		pos.x = pos.x - column*dx
+
+		-- find first spot not occupied by tech card
+		local vacant = false
+		while not vacant do
+			vacant = true
+			for _,cardOnTable in ipairs(techCardPositionsOnTable) do
+				cardPos = techBoard.positionToLocal(cardOnTable)
+				if math.floor(cardPos.x*10+0.5) == math.floor(pos.x*10+0.5) then
+					if math.floor(cardPos.z*10+0.5) == math.floor(pos.z*10+0.5) then
+						vacant = false
+						pos.z = pos.z + dz
+						break
+					end
+				end
+			end
+		end
+		return {position = techBoard.positionToWorld(pos), rotation = techBoard.getRotation()}
+	end
 
-		pos = snapPoints[pointorder[overrides[cardName]["upgrade"]]]
-		pos = factionSheet.positionToWorld(pos)
-	else
+	local function _getPosToTechDeck(cardName, techDeck)
 		-- place cards next to tech deck (take care on which side of the table)
-		pos = techDeck.getPosition()
+		local pos = techDeck.getPosition()
+		local rot = techDeck.getRotation()
+		local dz = 2
 		if pos.z > 0 then
 			pos.x = pos.x + 4
+			dz = -2
 		else
 			pos.x = pos.x - 4
 		end
+
+		-- check that the card is not moved over another existing tech card
```

Comment:
Could move this nearly identical "find an open spot" into a shared function that takes p0, dx, dz to find the first position = p0 + i * (dx, dz).

Reply:
Made a getFirstVacantPosition() in the technologyHelper. Might make more sense elsewhere, as this can potentially be a collision check for any object.